serverless/cmd/distribute/github/main.go

```diff
@@ -0,0 +1,305 @@
+
+	repo, err := github.NewRepository(ctx, opts.distributorRepo, *distributorBranch, opts.forkRepo, opts.gitUsername, opts.gitEmail, opts.githubAuthToken, opts.clonePath)
+	if err != nil {
+		glog.Exitf("Failed to set up repository: %v", err)
+	}
+
+	if err := distributeOnce(ctx, opts, repo); err != nil {
+		glog.Warningf("distributeOnce: %v", err)
+	}
+	if *interval > 0 {
+		for {
+			select {
+			case <-time.After(*interval):
+				glog.Infof("Wait time is up, going around again (%s)", *interval)
+			case <-ctx.Done():
+				return
+			}
+			if err := distributeOnce(ctx, opts, repo); err != nil {
+				glog.Warningf("distributeOnce: %v", err)
+			}
+		}
+	}
+}
+
+// distributeOnce a) polls the witness b) updates the fork c) proposes a PR if needed.
+func distributeOnce(ctx context.Context, opts *options, repo github.Repository) error {
+	// Pull forkRepo and get the ref for origin/master branch HEAD.
+	headRef, err := repo.PullAndGetHead()
+	if err != nil {
+		return err
+	}
+
+	// This will be used on both the witness and the distributor.
+	// At the moment the ID is arbitrary and is up to the discretion of the operators
+	// of these parties. We should address this. If we don't manage to do so in time,
+	// we'll need to allow this ID to be configured separately for each entity.
+	logID := opts.distConfig.Log.ID
+
+	wRaw, err := opts.witness.GetLatestCheckpoint(ctx, logID)
+	if err != nil {
+		return err
+	}
+	wCp, _, witnessNote, err := log.ParseCheckpoint(wRaw, opts.distConfig.Log.Origin, opts.logSigV, opts.witSigV)
+	if err != nil {
+		return fmt.Errorf("couldn't parse witnessed checkpoint: %v", err)
+	}
+	if nWitSigs, want := len(witnessNote.Sigs)-1, 1; nWitSigs != want {
+		return fmt.Errorf("checkpoint has %d witness sigs, want %d", nWitSigs, want)
+	}
+
+	// Now form a PR with the cosigned CP.
+	id := wcpID(*witnessNote)
+	branchName := fmt.Sprintf("refs/heads/witness_%s", id)
+	deleteBranch, err := repo.CreateLocalBranch(headRef, branchName)
+	if err != nil {
+		return fmt.Errorf("failed to create git branch for PR: %v", err)
+	}
+	defer deleteBranch()
+
+	outputPath := filepath.Join(opts.distributorPath, "logs", logID, "incoming", fmt.Sprintf("checkpoint_%s", id))
+	// First, check whether we've already managed to submit this CP into the incoming directory
+	if _, err := repo.ReadFile(outputPath); err == nil {
+		return fmt.Errorf("witnessed checkpoint already pending: %v", impl.ErrNoSignaturesAdded)
+	} else if !errors.Is(err, os.ErrNotExist) {
+		return fmt.Errorf("failed to check for existing pending checkpoint: %v", err)
+	}
+
+	msg := fmt.Sprintf("Witness checkpoint@%v", wCp.Size)
+	if err := repo.CommitFile(outputPath, wRaw, msg); err != nil {
+		return fmt.Errorf("failed to commit updated checkpoint.witnessed file: %v", err)
+	}
+
+	if err := repo.Push(); err != nil {
+		return fmt.Errorf("failed to push: %v", err)
+	}
+
+	glog.V(1).Info("Creating PR")
+	return repo.CreatePR(ctx, fmt.Sprintf("Witness @ %d", wCp.Size), branchName)
```

Comment:
nit/TODO: would be cool to add the witness' name to the PR title too.

Reply:
https://github.com/mhutchinson/mhutchinson-distributor/pull/4 - much better.